### PR TITLE
Refactor SQL in testcases and fix #174

### LINKED ISSUE #174: Error when remove tag from a plan's cases
Steps

1. Open a plan and go to Cases tab
1. Select menu on top of case list, Tag -> Remove Tag

```
Internal Server Error: /cases/tag/
Traceback (most recent call last):
  File "/home/vagrant/nitrate-env/lib/python2.7/site-packages/django/core/handlers/base.py", line 132, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/code/tcms/testcases/views.py", line 1700, in tag
    form.populate(case_ids=tcs)
  File "/code/tcms/testcases/forms.py", line 626, in populate
    testcase__in=case_ids).order_by('name').distinct()
  File "/home/vagrant/nitrate-env/lib/python2.7/site-packages/django/db/models/manager.py", line 127, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/home/vagrant/nitrate-env/lib/python2.7/site-packages/django/db/models/query.py", line 679, in filter
    return self._filter_or_exclude(False, *args, **kwargs)
  File "/home/vagrant/nitrate-env/lib/python2.7/site-packages/django/db/models/query.py", line 697, in _filter_or_exclude
    clone.query.add_q(Q(*args, **kwargs))
  File "/home/vagrant/nitrate-env/lib/python2.7/site-packages/django/db/models/sql/query.py", line 1310, in add_q
    clause, require_inner = self._add_q(where_part, self.used_aliases)
  File "/home/vagrant/nitrate-env/lib/python2.7/site-packages/django/db/models/sql/query.py", line 1338, in _add_q
    allow_joins=allow_joins, split_subq=split_subq,
  File "/home/vagrant/nitrate-env/lib/python2.7/site-packages/django/db/models/sql/query.py", line 1150, in build_filter
    lookups, parts, reffed_expression = self.solve_lookup_type(arg)
  File "/home/vagrant/nitrate-env/lib/python2.7/site-packages/django/db/models/sql/query.py", line 1036, in solve_lookup_type
    _, field, _, lookup_parts = self.names_to_path(lookup_splitted, self.get_meta())
  File "/home/vagrant/nitrate-env/lib/python2.7/site-packages/django/db/models/sql/query.py", line 1397, in names_to_path
    "Choices are: %s" % (name, ", ".join(available)))
FieldError: Cannot resolve keyword 'testcase' into field. Choices are: cases, id, name, testcasetag, testplan, testplantag, testrun, testruntag
[09/Apr/2017 11:13:45] "POST /cases/tag/ HTTP/1.1" 500 18058
```

### PR DESCRIPTION
**Pull request by atodorov**
This is a minor refactoring covered by existing tests.

The second commit fixes #174. I will see if I can add automated tests for it as well but please review.

### CODE REVIEW

**Comment by atodorov** [comment]
Note to self: the print and export SQLs are very similar to the ones used in `testplan/` so it is best they be refactored after #172 has been merged

File: tcms/testcases/models.py [hunk 1]
```diff
@@ -522,29 +520,27 @@ def mail(self, template, subject, context={}, to=[], request=None):
...
+
+        query.delete()
 
     def remove_component(self, component):
-        with transaction.atomic():
-            cursor = connection.writer_cursor
-            cursor.execute(SQL.REMOVE_COMPONENT, (self.case_id, component.id))
+        # note: cannot use self.component.remove(component) on a ManyToManyField
+        # which specifies an intermediary model so we use the model manager!
+        self.component.through.objects.filter(case=self.pk, component=component.pk).delete()
```
Comment (tkdchen) [inline tcms/testcases/models.py]: Use intermediary table can make it much easy to understand. 

```
TestCaseComponent.objects.filter(case=self, component=component).delete()
```

It's straightforward and no need of comment here.

Same comment to below changes.

File: tcms/testcases/views.py [hunk 2]
```diff
@@ -1696,7 +1697,7 @@ def tag(request):
                     return HttpResponse(json.dumps(ajax_response))
         return HttpResponse(json.dumps(ajax_response))
 
-    form = CaseTagForm(initial={'tag': request.REQUEST.get('o_tag')})
+    form = CaseTagForm(initial={'tag': request.POST.get('o_tag')})
```
Comment (tkdchen) [inline tcms/testcases/views.py]: These changes are related to the issue replacing ``request.REQUEST`` that I'm working on. I'll fix it later. I think you can remove these changes from this PR.

File: tcms/testcases/tests.py [hunk 3]
```diff
@@ -197,3 +198,33 @@ def test_remove_tag(self):
 
         tag_pks = list(self.case.tag.all().values_list('pk', flat=True))
         self.assertEqual([self.tag_fedora.pk], tag_pks)
+
+
+class CaseTagFormTest(BasePlanCase):
```
Comment (tkdchen) [inline tcms/testcases/tests.py]: Looks this test case does not need inherit from ``BasePlanCase``, as no data created in super ``setUpTestData`` is used.

File: tcms/testcases/forms.py [hunk 4]
```diff
@@ -622,7 +622,9 @@ class CaseTagForm(forms.Form):
 
     def populate(self, case_ids=None):
         if case_ids:
-            self.fields['o_tag'].queryset = TestTag.objects.filter(
-                testcase__in=case_ids).order_by('name').distinct()
+            tag_ids = TestCaseTag.objects.filter(
+                case__in=case_ids
+            ).values_list('tag').distinct()
+            self.fields['o_tag'].queryset = TestTag.objects.filter(pk__in=tag_ids).order_by('name')
```
Comment (tkdchen) [inline tcms/testcases/forms.py]: Replace ``testcase__in`` with ``cases__in`` works. No need to query ``TestCaseTag`` separately.

**Comment by tkdchen** [comment]
Please move ``Fix #174`` into commit message body and sign-off these two commits. Thanks.

File: tcms/testcases/views.py (hunk 2 above)
Comment (atodorov) [inline tcms/testcases/views.py]: This is why we use git, if we do conflict one of the PRs gets precedence (I'd say the one which was opened first) and the other one simply rebases and resolves the conflict once the first PR is merged.

I have split this change into a separate commit b/c it is independent of the other changes but I'm not going to remove it.

File: tcms/testcases/forms.py (hunk 4 above)
Comment (atodorov) [inline tcms/testcases/forms.py]: OK, this is counter intuitive but works. Django is able to follow backwards relationships filter. In this case the relationship is TestCaseTag -> TestTag but Django is able to figure out how to go backwards by using inner joins. 

IMO this is a really bad side of Django ORM because it may lead to unpredicted results. E.g. if you forget .distinct() at the end the results are different. Anyway I've added a comment in the code for future reference.

**Comment by atodorov** [comment]
Signed & rebased to latest upstream.

**Comment by tkdchen** [comment]
Merged. Thanks.
